scio-parquet/src/main/scala/com/spotify/scio/parquet/read/ParquetReadFn.scala

```diff
@@ -112,40 +130,51 @@ class ParquetReadFn[T, R](
         readSupport.prepareForRead(hadoopConf, fileMetadata, fileSchema, readContext)
       val columnIO = columnIOFactory.getColumnIO(readContext.getRequestedSchema, fileSchema, true)
 
-      var currentRowGroupIndex = tracker.currentRestriction.getFrom
-      (0L until currentRowGroupIndex).foreach(_ => reader.skipNextRowGroup())
-
-      while (tracker.tryClaim(currentRowGroupIndex)) {
-        val pages = reader.readNextRowGroup
-
-        logger.debug(
-          "row group {} read in memory. row count = {}",
-          currentRowGroupIndex,
-          pages.getRowCount
-        )
-
-        val recordReader = columnIO.getRecordReader(
-          pages,
-          recordConverter,
-          if (options.useRecordFilter) filter else FilterCompat.NOOP
-        )
-        readRowGroup(
-          currentRowGroupIndex,
-          pages.getRowCount,
-          file,
-          recordReader,
-          outputReceiver,
-          projectionFn
-        )
+      granularity match {
+        case File =>
+          val tryClaim = tracker.tryClaim(tracker.currentRestriction().getFrom)
+          var pages = reader.readNextRowGroup()
+          while (tryClaim && pages != null) {
+            try {
+              val recordReader = columnIO.getRecordReader(
+                pages,
+                recordConverter,
+                if (options.useRecordFilter) filter else FilterCompat.NOOP
+              )
+              readRowGroup(
+                0,
+                pages.getRowCount,
+                file,
+                recordReader,
+                outputReceiver,
+                projectionFn
+              )
+            } finally {
+              pages = reader.readNextRowGroup()
+            }
```

Comment:
Since the implementation is not really symmetric, we can probably go further and not check `tryClaim` every cycle.
What is the reason we wrap into a try in the `File` path compared to the `RowGroup` ?

Reply:
I added the `tryClaim` check because I noticed that without it, Dataflow would redundantly process files because multiple workers would be assigned the same piece of work... it actually duplicate-read like 10mil records without 😬 

the logic I really wanted was something Java-style like 

```java
if (!tracker.tryClaim(tracker.currentRestriction().getFrom)) {
  return;
}
while (...) {
  ..
}
```

but...scala 😅 

you're right the try/finally isn't necessary, I'll remove!